Freeze read only feature fixed for Jupyter 5.x (and compatible with Jupyter 4.x)

tnarik opened the pull request:
Proposed fix for #940.

jcb91 commented:
Thanks for this, it looks good to me. As noted in https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/885#issuecomment-280071268, there's also a `cell.metadata.deletable` attribute, though it seems that non-editable cells are implicitly non-deletable, so I guess we don't need to set that as well. However, it might be good to migrate & remove the old metadata values?

tnarik commented:
I wasn't sure about what you thought regarding that, but yes, a migration
of the metadata would be good to have for any notebook edited with an old
version of the plugin.

it's also pretty straightforward and I can do that in day or even sooner.

On 14 Apr 2017 8:36 p.m., "Josh Barnes" <notifications@github.com> wrote:

Thanks for this, it looks good to me. As noted in #885 (comment)
<https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/885#issuecomment-280071268>,
there's also a cell.metadata.deletable attribute, though it seems that
non-editable cells are implicitly non-deletable, so I guess we don't need
to set that as well. However, it might be good to migrate & remove the old
metadata values?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/958#issuecomment-294221073>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAGx4jxBkJrad1NerHjRdof2hOp1_Jjqks5rv8qhgaJpZM4M8avU>
.

tnarik commented:
Updated the notebook to read old format ('run_control>read_only') metadata.
Also fixed an issue by which, once touched by the plugin, the 'editable' metadata was set to 'false' (because in Jupyter 5.x `editable = true` means the flag won't be persisted to file).

jcb91 commented on src/jupyter_contrib_nbextensions/nbextensions/freeze/main.js:
```diff
@@ -116,9 +118,14 @@ define([
                 continue;
             }
             var state = 'normal';
+            // Old metadata format
```
Oh, I was thinking of something more like

```javascript
if (cell.metadata.run_control !== undefined && cell.metadata.run_control.read_only !== undefined) {
    if (cell.metadata.run_control.read_only) {
        cell.metadata.editable = false;
    }
    delete cell.metadata.run_control.read_only;
}
```

so that we adopt the new metadata just by loading the notebook - make sense?

tnarik commented:
I just didn't want to change much the existing code, even if refactoring
could make sense.

Opening the notebook will immediately take care of the "upgrade" because
loading the extension triggers the  initialization of the cell states and
the setting of the cell metadata (via `set_state`).

Maybe I'm missing your point, it's the first time I touch Jupyter extension
code.




On 18 Apr 2017 2:34 p.m., "Josh Barnes" <notifications@github.com> wrote:

*@jcb91* commented on this pull request.
------------------------------

In src/jupyter_contrib_nbextensions/nbextensions/freeze/main.js
<https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/958#discussion_r111954163>
:

> @@ -116,9 +118,14 @@ define([
                 continue;
             }
             var state = 'normal';
+            // Old metadata format

Oh, I was thinking of something more like

if (cell.metadata.run_control !== undefined &&
cell.metadata.run_control.read_only !== undefined) {
    if (cell.metadata.run_control.read_only) {
        cell.metadata.editable = false;
    }
    delete cell.metadata.run_control.read_only;
}

so that we adopt the new metadata just by loading the notebook - make sense?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/958#pullrequestreview-33230990>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAGx4v3d9ZFHGbKQgOwn5sxxrt0RDpQeks5rxLvYgaJpZM4M8avU>
.

jcb91 commented:
> I just didn't want to change much the existing code, even if refactoring could make sense.

Fair enough, that's a reasonable approach :smile: 

> Opening the notebook will immediately take care of the "upgrade" because
loading the extension triggers the  initialization of the cell states and
the setting of the cell metadata (via `set_state`).

Yes, it will work ok, in the sense that the new `metadata.editable` will be set correctly, __but__ it will also leave behind the old `metadata.run_control.read_only` value, which will override the `metadata.editable` each time the nbextension loads. I think it'd be neater to remove the old-style metadata attribute once it's been transferred to the new `editable` attribute. Does that make more sense?

tnarik commented:
Ah, sorry. Yes, it makes perfect sense. I thought I tested that but I didn't.

tnarik commented:
Created a new pull request, #968, instead of this one,
